### ISSUE TITLE
Instrumented applications crash when running in android phone/emulator after instrumentation .

### ISSUE THREAD

**Issue by maldil** [Sable/soot#834]
Hi everyone, 
I am in the process of android application instrumentation. The problem is the application is crashed when I open the application after successful installation in both emulator and phone. 
What I did was,
1. Changed the application with soot
2. Singed the application
3. Aligned the application
4. Installed in phone and the emulator
5. Open the application
Application doesn't responded.  

Debugging procedure
1. Removed the instrumentation code and did only the code transformation part.  (removed the code in `internalTransform(String s, Map<String, String> map)`).
2. Singed the application
3. Aligned the application
4. Installed and opened the application
Hence, we can conclude the application crashed even just with transformation (without modification).

Below is my instrumentation settings 
```
        PackManager.v().getPack("wjtp").add(new Transform("wjtp.myAnalysis", new APKDSceneTransformer(finalChangingMethods)));
        PackManager.v().runPacks();
        PackManager.v().writeOutput();
```
Below is the stack trace in log cat
```
25 14:26:46.960 1328-1613/? W/audio_hw_generic: Not supplying enough data to HAL, expected position 3127472 , only wrote 2991025
11-25 14:26:47.786 1626-2167/? I/ActivityManager: START u0 {act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=yogi.corporationapps.telescope.bigzoomhd/.HomeActivity (has extras)} from uid 10009 on display 0
11-25 14:26:47.797 1626-1642/? W/BroadcastQueue: Permission Denial: broadcasting Intent { act=com.android.launcher3.action.LAUNCH flg=0x10 (has extras) } from com.android.launcher3 (pid=2005, uid=10009) requires com.google.android.launcher.permission.RECEIVE_LAUNCH_BROADCASTS due to receiver com.google.android.gms/.icing.proxy.ApplicationLauncherReceiver
11-25 14:26:47.797 1626-1642/? W/BroadcastQueue: Permission Denial: receiving Intent { act=com.android.launcher3.action.LAUNCH flg=0x10 (has extras) } to com.google.android.gms/.chimera.GmsIntentOperationService$GmsExternalReceiver requires com.android.launcher3.permission.RECEIVE_LAUNCH_BROADCASTS due to sender com.android.launcher3 (uid 10009)
11-25 14:26:47.797 1626-1642/? W/BroadcastQueue: Permission Denial: broadcasting Intent { act=com.android.launcher3.action.LAUNCH flg=0x10 (has extras) } from com.android.launcher3 (pid=2005, uid=10009) requires com.google.android.launcher.permission.RECEIVE_LAUNCH_BROADCASTS due to receiver com.google.android.googlequicksearchbox/com.google.android.search.core.icingsync.ApplicationLaunchReceiver
11-25 14:26:47.813 1626-1780/? I/ActivityManager: Start proc 3512:yogi.corporationapps.telescope.bigzoomhd/u0a55 for activity yogi.corporationapps.telescope.bigzoomhd/.HomeActivity
11-25 14:26:47.829 1328-1613/? D/AudioFlinger: mixer(0xb4300000) throttle end: throttle time(11)
11-25 14:26:47.917 1626-1642/? W/art: Long monitor contention event with owner method=void com.android.server.am.ActivityManagerService.activityPaused(android.os.IBinder) from ActivityManagerService.java:6439 waiters=0 for 118ms
11-25 14:26:47.929 1236-1687/? D/gralloc_ranchu: gralloc_alloc: Creating ashmem region of size 3686400
11-25 14:26:47.952 3512-3512/? W/System: ClassLoader referenced unknown path: /data/app/yogi.corporationapps.telescope.bigzoomhd-1/lib/x86
11-25 14:26:47.970 3512-3512/? I/art: Rejecting re-init on previously-failed class java.lang.Class<yogi.corporationapps.telescope.bigzoomhd.HomeActivity>
11-25 14:26:47.970 3512-3512/? D/AndroidRuntime: Shutting down VM
11-25 14:26:47.970 3512-3512/? E/AndroidRuntime: FATAL EXCEPTION: main
                                                 Process: yogi.corporationapps.telescope.bigzoomhd, PID: 3512
                                                 java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{yogi.corporationapps.telescope.bigzoomhd/yogi.corporationapps.telescope.bigzoomhd.HomeActivity}: java.lang.ClassNotFoundException: Didn't find class "yogi.corporationapps.telescope.bigzoomhd.HomeActivity" on path: DexPathList[[zip file "/data/app/yogi.corporationapps.telescope.bigzoomhd-1/base.apk"],nativeLibraryDirectories=[/data/app/yogi.corporationapps.telescope.bigzoomhd-1/lib/x86, /vendor/lib, /system/lib]]
                                                     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2327)
                                                     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                     at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                     at android.os.Handler.dispatchMessage(Handler.java:102)
                                                     at android.os.Looper.loop(Looper.java:148)
                                                     at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                     at java.lang.reflect.Method.invoke(Native Method)
                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                  Caused by: java.lang.ClassNotFoundException: Didn't find class "yogi.corporationapps.telescope.bigzoomhd.HomeActivity" on path: DexPathList[[zip file "/data/app/yogi.corporationapps.telescope.bigzoomhd-1/base.apk"],nativeLibraryDirectories=[/data/app/yogi.corporationapps.telescope.bigzoomhd-1/lib/x86, /vendor/lib, /system/lib]]
                                                     at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
                                                     at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
                                                     at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
                                                     at android.app.Instrumentation.newActivity(Instrumentation.java:1067)
                                                     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2317)
                                                     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2476)
                                                     at android.app.ActivityThread.-wrap11(ActivityThread.java)
                                                     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1344)
                                                     at android.os.Handler.dispatchMessage(Handler.java:102)
                                                     at android.os.Looper.loop(Looper.java:148)
                                                     at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                     at java.lang.reflect.Method.invoke(Native Method)
                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                     Suppressed: java.lang.NoClassDefFoundError: yogi.corporationapps.telescope.bigzoomhd.HomeActivity
                                                     at dalvik.system.DexFile.defineClassNative(Native Method)
                                                     at dalvik.system.DexFile.defineClass(DexFile.java:226)
                                                     at dalvik.system.DexFile.loadClassBinaryName(DexFile.java:219)
                                                     at dalvik.system.DexPathList.findClass(DexPathList.java:338)
                                                     at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:54)
                                                             ... 13 more
                                                     Suppressed: java.lang.ClassNotFoundException: yogi.corporationapps.telescope.bigzoomhd.HomeActivity
                                                     at java.lang.Class.classForName(Native Method)
                                                     at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
                                                     at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
                                                     at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
                                                             ... 12 more
                                                  Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
11-25 14:26:47.971 1626-1778/? W/ActivityManager:   Force finishing activity yogi.corporationapps.telescope.bigzoomhd/.HomeActivity
```
The exception message was `java.lang.RuntimeException: Unable to instantiate activity ComponentInfo{yogi.corporationapps.telescope.bigzoomhd/yogi.corporationapps.telescope.bigzoomhd.HomeActivity}: java.lang.ClassNotFoundException: Didn't find class "yogi.corporationapps.telescope.bigzoomhd.HomeActivity"`

I would be really grateful if you can help me on this.

**Comment by maldil** [comment]
@mbenz89 with the latest commit we removed a validation right ? would this be a reason for this ? Application might have transformed without validation. However, this happens without any modification to the application.

**Comment by maldil** [comment]
@anddann @StevenArzt @mbenz89 
Could you please looking to this. I checked about this for several applications, transformed application (without modifications) are crashing in emulator/phone.  This was working before the recent clean/ build of my project.

**Comment by mbenz89** [comment]
Have you set Soots output to DEX before you running Soot? If not, Soot will not load excluded classes which are then missing during the write-out process.

Can you check if this class is in the original Apk? If you have checked this, I might have a look if you send me the Apk.

**Comment by maldil** [comment]
@mbenz89 Thank you for the response. I do updated the out put format as dex `Options.v().set_output_format(Options.output_format_dex);`. After few analysis, I identified the problem as, after constructing the call graph using `Soot.getAnalyzer().constructCallgraph();`  re-initialization of Soot should be done.  I am wondering Flowdroid changes the configurations when constructing the callgraph.  (First, I am constructing the call graph, and then do the instrumentation).

**Comment by mbenz89** [comment]
Flowdroid overwrites the Soot settings. You have to register your custom SootConfiguration to FlowDroid.

**Comment by maldil** [comment]
@mbenz89 yes of course.  you did, and thank you for that. However, It doesn't work for me. It seems, configuration are overwritten in somewhere apart from `public void setSootOptions(Options options)`.

**Comment by mbenz89** [comment]
Try this: options.set_include_all(true);

**Comment by maldil** [comment]
@mbenz89 I updated the configuration but the result was same. Soot has excluded some classes. However,  reinitialized soot (After constructing the call graph) transformed the  apk correctly. :O

**Comment by mbenz89** [comment]
yes, this is really because FlowDroid rewrites the config to exclude some things.
At my side it works well with this configuration: 
```
        SootConfigForAndroid sootConf = new SootConfigForAndroid() {
            @Override
            public void setSootOptions(Options options) {
                options.set_allow_phantom_refs(true);
                options.set_process_multiple_dex(true);
                // no exclusions here  to allow printing out the whole apk later
                options.set_include_all(true);
            }
        };
        setupApplication.setSootConfig(sootConf);
        setupApplication.constructCallGraph();
```
Note that I ommitted the `super.setSootOptions` since it contains excludes.

**Comment by maldil** [comment]
Oops. That might be the reason. I called the `super.setSootOptions`.
I guess, I may need to call the super class method, otherwise there might be difficulties when constructing the call graph (I assumed, but I'll check).

**Comment by mbenz89** [comment]
Were you able to resolve the issue?

**Comment by arvind4141** [comment]
@maldil any luck? I'm trying to instrument android app now but the output seems to be corrupted.

**Comment by maldil** [comment]
@mbenz89  thank you very much for your attention on this. I could resolve the problem by re-initializing the soot configuration after constructing the call graph. It seems there is a problem when instrumentation and call graph construction perform together in one code.  

@arvind4141 what do you mean by corruption ?

**Comment by arvind4141** [comment]
@maldil My newly instrumented application is deemed corrupted by the AndroidOS and I can't seem to run it. I followed the example that they have from the Wiki with very minor tweaks, but I believe its my configuration settings and code setup that's causing the issue. 

Do you mind sharing the part of your segment of your method code where the SootConfig / Instrumentation / and Callgraph generation is handled?

**Comment by arvind4141** [comment]
@maldil 

Here's my code segment. Please do kindly point me out if I'm doing anything wrong, thank you.
1. Technically I'm just trying to simulate the same instrumentation that is done from here:

> https://github.com/Sable/soot/wiki/Instrumenting-Android-Apps-with-Soot

Here's the code segment:

        public static void Instrument() {
        soot.G.reset();
        SootConfigForAndroid sootConf = new SootConfigForAndroid() {
            @Override
            public void setSootOptions(Options options) {
                super.setSootOptions(options);
                options.v().set_allow_phantom_refs(true);
                options.v().set_whole_program(true);
                options.v().set_prepend_classpath(true);
                options.v().set_validate(true);
                options.v().set_src_prec(Options.src_prec_apk);
                options.v().set_output_format(Options.output_format_dex);
                options.v().set_android_jars(LIB);
                options.v().set_process_dir(Collections.singletonList(APK));

            }
        };
        SetupApplication analyzer = new SetupApplication(LIB,APK);
        analyzer.setSootConfig(sootConf);
        analyzer.getConfig().setTaintAnalysisEnabled(false);
        analyzer.setCallbackFile(CB);
        Scene.v().loadNecessaryClasses();
        Scene.v().addBasicClass("java.io.PrintStream",SootClass.SIGNATURES);
        Scene.v().addBasicClass("java.lang.System", SootClass.SIGNATURES);

        PackManager.v().getPack("jtp").add(new Transform("jtp.myInstrumenter", new BodyTransformer() {

            @Override
            protected void internalTransform(final Body b, String phaseName, @SuppressWarnings("rawtypes") Map options) {
                final PatchingChain<Unit> units = b.getUnits();

                //important to use snapshotIterator here
                for(Iterator<Unit> iter = units.snapshotIterator(); iter.hasNext();) {
                    final Unit u = iter.next();
                    u.apply(new AbstractStmtSwitch() {

                        public void caseInvokeStmt(InvokeStmt stmt) {
                            InvokeExpr invokeExpr = stmt.getInvokeExpr();
                            if(invokeExpr.getMethod().getName().equals("onDraw")) {

                                Local tmpRef = addTmpRef(b);
                                Local tmpString = addTmpString(b);

                                // insert "tmpRef = java.lang.System.out;"
                                units.insertBefore(Jimple.v().newAssignStmt(
                                        tmpRef, Jimple.v().newStaticFieldRef(
                                                Scene.v().getField("<java.lang.System: java.io.PrintStream out>").makeRef())), u);

                                // insert "tmpLong = 'HELLO';"
                                units.insertBefore(Jimple.v().newAssignStmt(tmpString,
                                        StringConstant.v("HELLO")), u);

                                // insert "tmpRef.println(tmpString);"
                                SootMethod toCall = Scene.v().getSootClass("java.io.PrintStream").getMethod("void println(java.lang.String)");
                                units.insertBefore(Jimple.v().newInvokeStmt(
                                        Jimple.v().newVirtualInvokeExpr(tmpRef, toCall.makeRef(), tmpString)), u);

                                //check that we did not mess up the Jimple
                                b.validate();
                            }
                        }

                    });
                }
            }


        }));
        PackManager.v().runPacks();
        PackManager.v().writeOutput();
    }
    private static Local addTmpRef(Body body)
    {
        Local tmpRef = Jimple.v().newLocal("tmpRef", RefType.v("java.io.PrintStream"));
        body.getLocals().add(tmpRef);
        return tmpRef;
    }

    private static Local addTmpString(Body body)
    {
        Local tmpString = Jimple.v().newLocal("tmpString", RefType.v("java.lang.String"));
        body.getLocals().add(tmpString);
        return tmpString;
    }

Everything is sucessful when building , no errors , no exceptions. But the output APK is corrupted. And I strongly believe this is due to my configurations. Am I missing anything out? Thank you for the help once again!

**Comment by maldil** [comment]
@arvind4141 could you please do one thing before get into the code ? please comment the code inside the internalTransform and run the code. Then, try to install & run the transformed application.

**Comment by arvind4141** [comment]
@maldil I've just did a little experimenting and totally removed the transformer portion of my code.
Now it looks like this.
This is my objective from the below code.
1 - Get in a source APK.
2 - Output as dex (which is technically APK)
But it doesn't seem to work as well, the output file still remains corrupted.
Am i missing any key ingredients / code segment to make the output file work? Thank you for the help!

        public static void Instrument() {
        //prefer Android APK files// -src-prec apk
        soot.G.reset();
        SootConfigForAndroid sootConf = new SootConfigForAndroid() {
            @Override
            public void setSootOptions(Options options) {
                super.setSootOptions(options);
                options.v().set_allow_phantom_refs(true);
                options.v().set_whole_program(true);
                options.v().set_prepend_classpath(true);
                options.v().set_validate(true);
                options.v().set_src_prec(Options.src_prec_apk);
                options.v().set_output_format(Options.output_format_dex);
                options.v().set_android_jars(LIB);
                options.v().set_process_dir(Collections.singletonList(APK));
                options.set_include_all(true);
                options.set_process_multiple_dex(true);

            }
        };
        SetupApplication analyzer = new SetupApplication(LIB,APK);
        analyzer.setSootConfig(sootConf);
        analyzer.getConfig().setTaintAnalysisEnabled(false);
        analyzer.setCallbackFile(CB);
        analyzer.constructCallgraph();
        PackManager.v().runPacks();
        PackManager.v().writeOutput();
    }

**Comment by maldil** [comment]
I can't see the package details in this code. I strongly recommend you the paper https://link.springer.com/chapter/10.1007/978-3-642-40787-1_26 . Please go through it, if you haven't

**Comment by arvind4141** [comment]
@maldil Alrights thank you, sorry to bother you - but do you mind sharing with me your segment of code just for the **sootconfiguration** part and the **packmanager** part? I could use that as a reference start point. I would really appreciate it if its possible, thank you!

**Comment by maldil** [comment]
@arvind4141  I'hv implemented according to the details in the paper. Please go through it.

**Comment by arvind4141** [comment]
@maldil Hello, it seems that the issue was due to the API level being set. For some reason setting the API level more than > 23 will cause a "corrupted" DEX file to be output. 
This can be mitigated by using the code below:

     Options.v().set_android_api_version(23);

@maldil So it seems that as suspected my soot configuration had some issue to it and it was all due to this API level. Now the output DEX file is perfectly fine. I've resolved the issue, thank you so much friend!
